<a href="https://colab.research.google.com/github/gettokeroro/DADS5001_MINI_PROJECT/blob/main/Job_Application_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [29]:
# Install dependencies as needed:
# pip install kagglehub[pandas-datasets]
import kagglehub
from kagglehub import KaggleDatasetAdapter

# Set the path to the file you'd like to load (relative path within the dataset)
file_path = "postings.csv"

# Load the latest version
df = kagglehub.dataset_load(
  KaggleDatasetAdapter.PANDAS,
  "arshkon/linkedin-job-postings",
  file_path,
  # Provide any additional arguments like
  # sql_query or pandas_kwargs. See the
  # documenation for more information:
  # https://github.com/Kaggle/kagglehub/blob/main/README.md#kaggledatasetadapterpandas
)

display(df.head())

Using Colab cache for faster access to the 'linkedin-job-postings' dataset.


,job_id,company_name,title,description,max_salary,pay_period,location,company_id,views,med_salary,...,skills_desc,listed_time,posting_domain,sponsored,work_type,currency,compensation_type,normalized_salary,zip_code,fips
0,921716,Corcoran Sawyer Smith,Marketing Coordinator,Job descriptionA leading real estate firm in N...,20.0,HOURLY,"Princeton, NJ",2774458.0,20.0,NaN,...,Requirements: \n\nWe are seeking a College or ...,1.713398e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,38480.0,8540.0,34021.0
1,1829192,NaN,Mental Health Therapist/Counselor,"At Aspen Therapy and Wellness , we are committ...",50.0,HOURLY,"Fort Collins, CO",NaN,1.0,NaN,...,NaN,1.712858e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,83200.0,80521.0,8069.0
2,10998357,The National Exemplar,Assitant Restaurant Manager,The National Exemplar is accepting application...,65000.0,YEARLY,"Cincinnati, OH",64896719.0,8.0,NaN,...,We are currently accepting resumes for FOH - A...,1.713278e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,55000.0,45202.0,39061.0
3,23221523,"Abrams Fensterman, LLP",Senior Elder Law / Trusts and Estates Associat...,Senior Associate Attorney - Elder Law / Trusts...,175000.0,YEARLY,"New Hyde Park, NY",766262.0,16.0,NaN,...,This position requires a baseline understandin...,1.712896e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,157500.0,11040.0,36059.0
4,35982263,NaN,Service Technician,Looking for HVAC service tech with experience ...,80000.0,YEARLY,"Burlington, IA",NaN,3.0,NaN,...,NaN,1.713452e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,70000.0,52601.0,19057.0


In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 122132 entries, 0 to 123848
Data columns (total 31 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   job_id                      122132 non-null  int64  
 1   company_name                122130 non-null  object 
 2   title                       122132 non-null  object 
 3   description                 122126 non-null  object 
 4   max_salary                  29339 non-null   float64
 5   pay_period                  35564 non-null   object 
 6   location                    122132 non-null  object 
 7   company_id                  122132 non-null  float64
 8   views                       120505 non-null  float64
 9   med_salary                  6225 non-null    float64
 10  min_salary                  29339 non-null   float64
 11  formatted_work_type         122132 non-null  object 
 12  applies                     22896 non-null   float64
 13  original_listed_tim

In [31]:
import kagglehub
from kagglehub import KaggleDatasetAdapter
import pandas as pd
import os

# Base path for the dataset files
DATASET_PATH = "/kaggle/input/linkedin-job-postings/"

# Dictionary to map file names to their respective subdirectories
file_subdirs = {
    "companies.csv": "companies/",
    "company_industries.csv": "companies/",
    "employee_counts.csv": "companies/",
    "industries.csv": "mappings/",
    "skills.csv": "mappings/",
    "job_skills.csv": "jobs/",
    "job_industries.csv": "jobs/"
}

# List of additional files to load (keys of the dictionary)
additional_files = list(file_subdirs.keys())

# Load each file into a separate DataFrame
for file_name in additional_files:
    print(f"\nLoading {file_name}...")
    try:
        # Create a meaningful DataFrame name (e.g., df_companies)
        df_name = file_name.replace('.csv', '')
        # Construct the full path using the subdirectory information
        full_file_path = os.path.join(DATASET_PATH, file_subdirs[file_name], file_name)
        globals()[f'df_{df_name}'] = pd.read_csv(full_file_path)
        print(f"First 5 records of df_{df_name}:")
        display(globals()[f'df_{df_name}'].head())
    except Exception as e:
        print(f"Error loading {file_name}: {e}")

# You can now access these DataFrames using names like df_companies, df_industries, etc.


Loading companies.csv...
First 5 records of df_companies:


,company_id,name,description,company_size,state,country,city,zip_code,address,url
0,1009,IBM,"At IBM, we do more than work. We create. We cr...",7.0,NY,US,"Armonk, New York",10504,International Business Machines Corp.,https://www.linkedin.com/company/ibm
1,1016,GE HealthCare,Every day millions of people feel the impact o...,7.0,0,US,Chicago,0,-,https://www.linkedin.com/company/gehealthcare
2,1025,Hewlett Packard Enterprise,Official LinkedIn of Hewlett Packard Enterpris...,7.0,Texas,US,Houston,77389,1701 E Mossy Oaks Rd Spring,https://www.linkedin.com/company/hewlett-packa...
3,1028,Oracle,We’re a cloud technology company that provides...,7.0,Texas,US,Austin,78741,2300 Oracle Way,https://www.linkedin.com/company/oracle
4,1033,Accenture,Accenture is a leading global professional ser...,7.0,0,IE,Dublin 2,0,Grand Canal Harbour,https://www.linkedin.com/company/accenture



Loading company_industries.csv...
First 5 records of df_company_industries:


,company_id,industry
0,391906,Book and Periodical Publishing
1,22292832,Construction
2,20300,Banking
3,3570660,Book and Periodical Publishing
4,878353,Staffing and Recruiting



Loading employee_counts.csv...
First 5 records of df_employee_counts:


,company_id,employee_count,follower_count,time_recorded
0,391906,186,32508,1712346173
1,22292832,311,4471,1712346173
2,20300,1053,6554,1712346173
3,3570660,383,35241,1712346173
4,878353,52,26397,1712346173



Loading industries.csv...
First 5 records of df_industries:


,industry_id,industry_name
0,1,Defense and Space Manufacturing
1,3,Computer Hardware Manufacturing
2,4,Software Development
3,5,Computer Networking Products
4,6,"Technology, Information and Internet"



Loading skills.csv...
First 5 records of df_skills:


,skill_abr,skill_name
0,ART,Art/Creative
1,DSGN,Design
2,ADVR,Advertising
3,PRDM,Product Management
4,DIST,Distribution



Loading job_skills.csv...
First 5 records of df_job_skills:


,job_id,skill_abr
0,3884428798,MRKT
1,3884428798,PR
2,3884428798,WRT
3,3887473071,SALE
4,3887465684,FIN



Loading job_industries.csv...
First 5 records of df_job_industries:


,job_id,industry_id
0,3884428798,82
1,3887473071,48
2,3887465684,41
3,3887467939,82
4,3887467939,80


# Prepare Data  & Cleansing

**1. Remove rows with NaN `company_id`**

**2. Convert `original_listed_time` and `expiry` to date format**


**3. Replace NaN with 0 in `remote_allowed`**


**4. Merge `df_job_industries` with `df_industries`**  


**5. Merge `postings` with `df_job_industries_with_names`**  


**6. The `time_record period` is too short, so I'll just use the latest row to look at the company size.**


### 1. Remove rows with NaN **company_id**

#### Dropped null values in company_id to improve job application trend analysis, as unidentified companies would skew the data.
---

In [ ]:
df.dropna(subset=['company_id'], inplace=True)

### 1. Convert original_listed_time and expiry to date format

---



In [42]:
selected_columns_df = df[['original_listed_time', 'expiry']]
display(selected_columns_df.head())

,original_listed_time,expiry
0,1.713398e+12,1.715990e+12
2,1.713278e+12,1.715870e+12
3,1.712896e+12,1.715488e+12
5,1.713456e+12,1.716048e+12
6,1.712861e+12,1.715453e+12


In [41]:
selected_columns_df['original_listed_time_dt'] = pd.to_datetime(selected_columns_df['original_listed_time'], unit='ms')
selected_columns_df['expiry_dt'] = pd.to_datetime(selected_columns_df['expiry'], unit='ms')

print("DataFrame with original and converted date columns:")
display(selected_columns_df.head())

df['original_listed_time'] = pd.to_datetime(df['original_listed_time'], unit='ms')
df['expiry'] = pd.to_datetime(df['expiry'], unit='ms')

DataFrame with original and converted date columns:


/tmp/ipython-input-3169040767.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  selected_columns_df['original_listed_time_dt'] = pd.to_datetime(selected_columns_df['original_listed_time'], unit='ms')
/tmp/ipython-input-3169040767.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  selected_columns_df['expiry_dt'] = pd.to_datetime(selected_columns_df['expiry'], unit='ms')


,original_listed_time,expiry,original_listed_time_dt,expiry_dt
0,1.713398e+12,1.715990e+12,2024-04-17 23:45:08,2024-05-17 23:45:08
2,1.713278e+12,1.715870e+12,2024-04-16 14:26:54,2024-05-16 14:26:54
3,1.712896e+12,1.715488e+12,2024-04-12 04:23:32,2024-05-12 04:23:32
5,1.713456e+12,1.716048e+12,2024-04-18 16:01:39,2024-05-18 16:01:39
6,1.712861e+12,1.715453e+12,2024-04-11 18:43:39,2024-05-11 18:43:39


###2. Replace NaN with 0 in remote_allowed

---

In [37]:
print("Unique values in 'remote_allowed' column:")
print(df['remote_allowed'].unique())

Unique values in 'remote_allowed' column:
[nan  1.]


In [38]:
df['remote_allowed'].fillna(0, inplace=True)
print(df['remote_allowed'].unique())

[0. 1.]


/tmp/ipython-input-3893761681.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['remote_allowed'].fillna(0, inplace=True)


### 3. Merge Job Industries with Industry Names

---

In [32]:
df_job_industries_with_names = pd.merge(
    df_job_industries,
    df_industries,
    on='industry_id',
    how='left'
)

print("First 5 records of df_job_industries_with_names:")
display(df_job_industries_with_names.head())

First 5 records of df_job_industries_with_names:


,job_id,industry_id,industry_name
0,3884428798,82,Book and Periodical Publishing
1,3887473071,48,Construction
2,3887465684,41,Banking
3,3887467939,82,Book and Periodical Publishing
4,3887467939,80,Advertising Services


### 4. Merge Job Postings with Industries


In [33]:
df_merged = pd.merge(df, df_job_industries_with_names, on='job_id', how='left')

print("First 5 rows of the merged DataFrame with industry names:")
display(df_merged.head())


First 5 rows of the merged DataFrame with industry names:


,job_id,company_name,title,description,max_salary,pay_period,location,company_id,views,med_salary,...,posting_domain,sponsored,work_type,currency,compensation_type,normalized_salary,zip_code,fips,industry_id,industry_name
0,921716,Corcoran Sawyer Smith,Marketing Coordinator,Job descriptionA leading real estate firm in N...,20.0,HOURLY,"Princeton, NJ",2774458.0,20.0,NaN,...,NaN,0,FULL_TIME,USD,BASE_SALARY,38480.0,8540.0,34021.0,44.0,Real Estate
1,1829192,NaN,Mental Health Therapist/Counselor,"At Aspen Therapy and Wellness , we are committ...",50.0,HOURLY,"Fort Collins, CO",NaN,1.0,NaN,...,NaN,0,FULL_TIME,USD,BASE_SALARY,83200.0,80521.0,8069.0,NaN,NaN
2,10998357,The National Exemplar,Assitant Restaurant Manager,The National Exemplar is accepting application...,65000.0,YEARLY,"Cincinnati, OH",64896719.0,8.0,NaN,...,NaN,0,FULL_TIME,USD,BASE_SALARY,55000.0,45202.0,39061.0,32.0,Restaurants
3,23221523,"Abrams Fensterman, LLP",Senior Elder Law / Trusts and Estates Associat...,Senior Associate Attorney - Elder Law / Trusts...,175000.0,YEARLY,"New Hyde Park, NY",766262.0,16.0,NaN,...,NaN,0,FULL_TIME,USD,BASE_SALARY,157500.0,11040.0,36059.0,9.0,Law Practice
4,35982263,NaN,Service Technician,Looking for HVAC service tech with experience ...,80000.0,YEARLY,"Burlington, IA",NaN,3.0,NaN,...,NaN,0,FULL_TIME,USD,BASE_SALARY,70000.0,52601.0,19057.0,122.0,Facilities Services


In [44]:
del df_industries
del df_job_industries_with_names

# 5. The most up-to-date information for each company

In [45]:
df_employee_counts['time_recorded'] = pd.to_datetime(df_employee_counts['time_recorded'], unit='s')

# Sort by company_id and time_recorded to ensure the latest record is kept
df_employee_counts_cleaned = df_employee_counts.sort_values(by=['company_id', 'time_recorded'])

# Drop duplicates based on company_id, keeping the last (which is the one with max time_recorded)
df_employee_counts_cleaned = df_employee_counts_cleaned.drop_duplicates(subset='company_id', keep='last')

print("First 5 records of df_employee_counts after cleaning (one row per company_id based on max time_recorded):")
display(df_employee_counts_cleaned.head(20))

First 5 records of df_employee_counts after cleaning (one row per company_id based on max time_recorded):


,company_id,employee_count,follower_count,time_recorded
30056,1009,311223,16314846,2024-04-19 04:34:15
35604,1016,57001,2196350,2024-04-19 23:23:51
21766,1025,79559,3588329,2024-04-18 15:14:18
32109,1028,191374,9497909,2024-04-19 15:21:13
35631,1033,565191,11890321,2024-04-19 23:30:51
32107,1035,227317,22253257,2024-04-19 15:21:13
33665,1038,435078,15221049,2024-04-19 21:18:03
22303,1043,217641,6814908,2024-04-18 16:24:32
22629,1044,291883,6085704,2024-04-18 16:45:10
32426,1052,174692,1485885,2024-04-19 18:14:17


### Cleansing
1. Remove rows with NaN **company_id**  
**Reasoning:** To ensure data integrity, I'll remove rows where the `company_id` is missing (NaN), as a company ID is crucial for identification and linking data. The `df.info()` output showed that `company_id` has non-null values, so removing NaN will clean this column.

2. Replace NaN with 0 in **remote_allowed**  
**Reasoning:** The `remote_allowed` column indicates if remote work is allowed. Replacing NaN with 0 assumes that if it's not explicitly stated as allowed (i.e., not 1), then it's not allowed. This converts missing values into a meaningful category.

## Summary:

### Data Analysis Key Findings

*   The `df_job_industries_with_names` DataFrame was successfully created by merging `df_job_industries` and `df_industries` on `industry_id`, linking `job_id` to both `industry_id` and `industry_name`. For example, `industry_id` 82 was mapped to "Book and Periodical Publishing" and `industry_id` 48 to "Construction".
*   The main job postings DataFrame (`df`) was successfully merged with `df_job_industries_with_names` to create `df_merged`, adding `industry_id` and `industry_name` columns.
*   The `df_merged` DataFrame contains 160,399 entries and 33 columns. After the final merge, 158,963 job postings have an associated `industry_id`, and 158,863 have an `industry_name`. This indicates that approximately 99.1% of job postings were successfully enriched with industry information, while a small fraction (around 0.9%) did not have a matching industry. For instance, `job_id` 921716 is now associated with `industry_id` 44.0 and `industry_name` "Real Estate".
*   The `df_employee_counts` DataFrame was cleaned to `df_employee_counts_cleaned`, ensuring that each `company_id` has only one record, corresponding to the latest `time_recorded`.
*   For the main `df` DataFrame:
    *   Rows with NaN values in `company_id` were removed, reducing the DataFrame from 123,849 to 122,132 entries.
    *   NaN values in the `remote_allowed` column were successfully replaced with 0.

### Next Steps

*   The `df_merged` DataFrame is now prepared for further analysis, allowing for examination of job postings by their respective industries.
*   The `df_employee_counts_cleaned` DataFrame is ready to be used for company-level analysis based on the latest employee counts.
*   You might want to further investigate the approximately 0.9% of job postings in `df_merged` that did not receive industry information to understand if these are missing data, unclassified industries, or require a different matching strategy.
*   Consider further handling of other missing values in `df_merged` (e.g., `max_salary`, `min_salary`, `skills_desc`, `posting_domain`).

### Cleansing Task 2: Replace NaN with 0 in `remote_allowed`

**Reasoning:** The `remote_allowed` column indicates if remote work is allowed. Replacing NaN with 0 assumes that if it's not explicitly stated as allowed (i.e., not 1), then it's not allowed. This converts missing values into a meaningful category.